programs/bpf_loader/src/syscalls.rs

```diff
@@ -2299,126 +2516,1803 @@ fn call<'a>(
+                }
+                return;
+            }
+        };
+        let mut bn_ctx = match BigNumContext::new() {
+            Ok(bn) => bn,
+            Err(es) => {
+                if let Some(reason) = es.errors()[0].reason() {
+                    *result = Err(SyscallError::BigNumberOpenSSLError(reason.to_string()).into());
+                } else {
+                    *result = Err(SyscallError::BigNumberOpenSSLErrorUnknown.into());
+                }
+                return;
+            }
+        };
+        let bn_syscall_result = question_mark!(
+            translate_type_mut::<FfiBigNumber>(
+                memory_mapping,
+                bn_ffi_out_addr,
+                self.loader_id,
+                true
+            ),
+            result
+        );
+
+        if bn_syscall_result.data_len > MAX_BIGNUM_BYTE_SIZE {
+            *result = Err(SyscallError::BigNumberResultSizeError.into());
+        } else {
+            match bn_result.mod_exp(&lhs_bn, &rhs_bn, &mod_bn, &mut bn_ctx) {
+                Ok(_) => bignum_to_ffibignumber(
+                    &bn_result,
+                    bn_syscall_result,
+                    self.loader_id,
+                    memory_mapping,
+                    result,
+                ),
+                Err(e) => {
+                    if let Some(reason) = e.errors()[0].reason() {
+                        *result = Err(SyscallError::BigNumberModExpError(reason.to_string()).into())
+                    } else {
+                        *result = Err(SyscallError::BigNumberOpenSSLErrorUnknown.into())
+                    }
+                }
+            }
+        }
+    }
+}
+/// BIGNUM sol_bignum_mod_mul
+struct SyscallBigNumModMul<'a> {
+    cost: u64,
+    word_cost: u64,
+    word_div_cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
+    loader_id: &'a Pubkey,
+}
+impl<'a> SyscallObject<BpfError> for SyscallBigNumModMul<'a> {
+    fn call(
+        &mut self,
+        lhs_ffi_in_addr: u64,
+        rhs_ffi_in_addr: u64,
+        mod_ffi_in_addr: u64,
+        bn_ffi_out_addr: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        // Get the LHS FfiBigNumber value structure
+        let bn_syscall_lhs = question_mark!(
+            translate_type::<FfiBigNumber>(memory_mapping, lhs_ffi_in_addr, self.loader_id, true),
+            result
+        );
+        // Get the RHS FfiBigNumber value structure
+        let bn_syscall_rhs = question_mark!(
+            translate_type::<FfiBigNumber>(memory_mapping, rhs_ffi_in_addr, self.loader_id, true),
+            result
+        );
+        // Get the MOD FfiBigNumber value structure
+        let bn_syscall_mod = question_mark!(
+            translate_type::<FfiBigNumber>(memory_mapping, mod_ffi_in_addr, self.loader_id, true),
+            result
+        );
+        // Compute costs
+        // Compute costs
```

Comment:
Duplicate line

Reply:
Pushed in d39ebb80f035f15af1591ebb65f242ba2ea43053